test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/KeepAliveTimeoutTests.cs

```diff
@@ -0,0 +1,223 @@
+                {
+                    await connection.Send(
+                        "POST / HTTP/1.1",
+                        "Content-Length: 8",
+                        "",
+                        "a");
+                    await Task.Delay(LongDelay);
+                    await connection.Send("bcdefgh");
+                    await ReceiveResponse(connection);
+                }
+            }
+        }
+
+        [Fact]
+        public async Task KeepAliveTimeoutNotTriggeredMidChunkedRequest()
+        {
+            using (var host = StartHost())
+            {
+                using (var connection = new TestConnection(host.GetPort()))
+                {
+                    await connection.Send(
+                            "POST / HTTP/1.1",
+                            "Transfer-Encoding: chunked",
+                            "",
+                            "5", "hello",
+                            "");
+                    await Task.Delay(LongDelay);
+                    await connection.Send(
+                            "6", " world",
+                            "0",
+                             "",
+                             "");
+                    await ReceiveResponse(connection);
+                }
+            }
+        }
+
+        [Fact]
+        public async Task ConnectionTimesOutWhenOpenedButNoRequestSent()
+        {
+            using (var host = StartHost())
+            {
+                using (var connection = new TestConnection(host.GetPort()))
+                {
+                    await Task.Delay(LongDelay);
+                    await Assert.ThrowsAsync<IOException>(async () =>
+                    {
+                        await connection.Send(
+                            "GET / HTTP/1.1",
+                            "",
+                            "");
+                    });
+                }
+            }
+        }
+
+        private static IWebHost StartHost()
+        {
+            var host = new WebHostBuilder()
+                .UseKestrel(options =>
+                {
+                    options.AddServerHeader = false;
+                    options.Limits.KeepAliveTimeout = KeepAliveTimeout;
+                })
+                .UseUrls("http://127.0.0.1:0/")
+                .Configure(app =>
+                {
+                    app.Run(async context =>
+                    {
+                        const string response = "hello, world";
+                        context.Response.ContentLength = response.Length;
+                        await context.Response.WriteAsync(response);
+                    });
+                })
+                .Build();
+
+            host.Start();
+            return host;
+        }
+
+        private async Task ReceiveResponse(TestConnection connection)
+        {
+            await connection.Receive(
+                "HTTP/1.1 200 OK",
+                "");
+            await connection.ReceiveStartsWith("Date: ");
```

Comment:
Use a `MockSystemClock` like `TestServer`. It was nice to get rid of `ReceiveStartsWith`, and I don't want to see it back.